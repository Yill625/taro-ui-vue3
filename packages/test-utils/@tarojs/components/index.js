export const Text = 'text'
export const Image = 'image'
export const Icon = 'icon'
export const ScrollView = 'scroll-view'
export const View = 'view'
export const RichText = 'rich-text'
export const Button = 'button'
export const CheckboxGroup = 'checkbox-group'
export const Editor = 'editor'
export const Form = 'form'
export const Label = 'label'
export const PickerView = 'picker-view'
export const PickerViewColumn = 'picker-view-column'
export const CoverImage = 'cover-image'
export const CoverView = 'cover-view'
export const MoveableArea = 'moveable-area'
export const MoveableView = 'moveable-view'
export const Swiper = 'swiper'
export const FunctionalPageNavigator = 'functional-page-navigator'
export const Navigator = 'navigator'
export const Audio = 'audio'
export const Camera = 'camera'
export const LivePlayer = 'live-player'
export const Map = 'map'
export const Ad = 'ad'
export const OfficialAccount = 'official-account'
export const OpenData = 'open-data'
export const WebView = 'web-view'
export const NavigationBar = 'navigation-bar'
export const Block = 'block'
export const Canvas = 'canvas'
export const Checkbox = 'checkbox'
export const Progress = 'progress'
export const RadioGroup = 'radio-group'
export const Radio = 'radio'
export const SwiperItem = 'swiper-item'
export const Video = 'video'
export const Input = 'input'
export const Textarea = 'textarea'
export const Picker = 'picker'
export const Switch = 'switch'
export const Slider = 'slider'