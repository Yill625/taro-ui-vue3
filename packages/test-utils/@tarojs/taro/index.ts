
const ENV_TYPE = {
  WEAPP: 'WEAPP',
  WEB: 'WEB',
  RN: 'RN',
  SWAN: 'SWAN',
  ALIPAY: 'ALIPAY',
  TT: 'TT',
  QQ: 'QQ',
  JD: 'JD',
  QUICKAPP: 'QUICKAPP'
}

const envs = {
  "weapp": ENV_TYPE.WEAPP,
  "swan": ENV_TYPE.SWAN,
  "alipay": ENV_TYPE.ALIPAY,
  "h5": ENV_TYPE.WEB,
  "tt": ENV_TYPE.TT,
  "quickapp": ENV_TYPE.QUICKAPP,
  "qq": ENV_TYPE.QQ,
  "jd": ENV_TYPE.JD,
}

export const getEnvMockFn = () => {
  if (!process.env.TARO_ENV) {
    process.env.TARO_ENV = 'h5'
  }
  return envs[process.env.TARO_ENV]
}

export const pxTransformMockFn = (size, designWidth = 750) => {
  if (!process.env.TARO_ENV) {
    process.env.TARO_ENV = 'h5'
  }

  const deviceRatio = {
    640: 2.34 / 2,
    750: 1,
    828: 1.81 / 2
  }

  if (!size) return ''

  return process.env.TARO_ENV === 'h5'
    ? Math.ceil(parseInt(`${size}`, 10) / 40 * 640 / designWidth * 10000) / 10000 + 'rem'
    : parseInt(size, 10) * deviceRatio[designWidth] + 'rpx'
}

export const querySelectorMockFn = (
  rect0 = { width: 30, left: 30, height: 30, right: 30 }
) => (): any => {
  const query = {
    exec: (cb) => { cb([rect0]) }
  }
  query['select'] = jest.fn().mockReturnValue({
    boundingClientRect: jest.fn().mockReturnValue(query)
  })

  return query
}

export const chooseImageMockFn = () => {
  return new Promise<any>((resolve) => {
    resolve({
      tempFilePaths: ['/path/to/file'],
      tempFiles: [{ path: '/path/to/file', size: 256 }],
      errMsg: ''
    })
  })
}

document.querySelector = jest.fn((sel) => {
  return {
    getBoundingClientRect: jest.fn().mockReturnValueOnce({ width: 30 })
  }
})