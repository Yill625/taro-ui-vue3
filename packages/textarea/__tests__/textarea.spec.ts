import {
  genMountFn,
  testPropClassAndStyle
} from '@taro-ui-vue3/test-utils/helper'
import { pxTransformMockFn } from '@taro-ui-vue3/test-utils/@tarojs/taro'
import { hyphenate, capitalize } from '@vue/shared'
import AtTextarea from '../index.vue'
import Taro from '@tarojs/taro'

Taro.pxTransform = pxTransformMockFn

const mountFn = genMountFn(AtTextarea)

describe('AtTextarea', () => {
  it('should render all nodes and match snapshot', () => {
    const wrapper = mountFn()
    expect(wrapper.element).toMatchSnapshot()
  })

  testPropClassAndStyle(mountFn)

  it.each([
    'cursorSpacing', 'maxLength'
  ])('should render prop %s', (propName) => {
    const wrapper = mountFn({ [propName]: 300 })
    const attrName = propName === 'maxLength'
      ? propName.toLowerCase()
      : hyphenate(propName)

    expect(
      wrapper
        .find('.at-textarea__textarea')
        .attributes(attrName)
    ).toBe('300')

    if (propName === 'maxLength') {
      expect(
        wrapper
          .find('.at-textarea__counter')
          .text()
      ).toBe('0 / 300')
    }
  })

  it('should render maxLength + 500 more characters if allows text to overflow', () => {
    const value = 'maxLength'.repeat(50)
    const wrapper = mountFn({
      modelValue: value,
      maxLength: 300,
      textOverflowForbidden: false
    })

    expect(
      wrapper
        .find('.at-textarea--error')
        .exists()
    ).toBeTruthy()

    expect(
      wrapper
        .find('.at-textarea__textarea')
        .attributes('maxlength')
    ).toBe('800')

    expect(
      wrapper
        .find('.at-textarea__counter')
        .text()
    ).toBe('450 / 300')
  })

  it.each([
    ['placeholder', 'test'],
    ['placeholderClass', 'test'],
    ['placeholderStyle', 'color: red;'],
  ])('should render prop %s', (propName, expected) => {
    const wrapper = mountFn({ [`${propName}`]: expected })
    expect(
      wrapper
        .find('.at-textarea__textarea')
        .attributes(hyphenate(propName))
    ).toContain(expected)
  })

  it.each([
    'autoFocus',
    'fixed',
    'showConfirmBar',
    // 'disabled',
    // 'focus'
  ])('should render prop %s', async (propName) => {
    const wrapper = mountFn({ [`${propName}`]: true })
    await wrapper.vm.$nextTick()
    expect(
      wrapper
        .find('.at-textarea__textarea')
        .attributes(hyphenate(propName))
    ).toBe('true')
  })

  it('should render prop count', async () => {
    const wrapper = mountFn()

    const countNode = wrapper.find('.at-textarea__counter')
    expect(countNode.text()).toBe('0 / 200')

    await wrapper.setProps({ count: false })
    expect(
      wrapper
        .find('.at-textarea__counter')
        .exists()
    ).toBeFalsy()
  })

  it.skip('should render prop height', () => {
    const wrapper = mountFn({ height: 3000 })
    expect(
      wrapper
        .find('.at-textarea__textarea')
        .attributes('style')
    ).toBe(`height: ${pxTransformMockFn(3000)};`)
  })
})

describe('AtTextArea events', () => {
  it.each([
    'focus',
    'blur',
    'confirm',
    'linechange',
  ])('should emit %s', async (eventName) => {
    const eventHandler = jest.fn()
    const wrapper = mountFn({ [`on${capitalize(eventName)}`]: eventHandler })

    await wrapper
      .find('.at-textarea__textarea')
      .trigger(eventName, { value: 'test' })

    expect(wrapper.emitted()).toHaveProperty(eventName)
    expect(eventHandler).toBeCalled()
  })

  it('should emit update:modelValue if using v-model', async () => {
    const eventHandler = jest.fn()
    const wrapper = mountFn({
      'onUpdate:modelValue': eventHandler
    })

    await wrapper
      .find('.at-textarea__textarea')
      .trigger('input', { detail: { value: 'test' } })

    expect(wrapper.emitted()).toHaveProperty('update:modelValue')
    expect(eventHandler).toBeCalled()
  })
})
