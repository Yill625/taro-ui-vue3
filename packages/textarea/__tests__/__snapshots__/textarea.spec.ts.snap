// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AtTextarea should render all nodes and match snapshot 1`] = `
<view
  class="at-textarea at-textarea--WEB"
>
  <textarea
    auto-focus="false"
    class="at-textarea__textarea"
    cursor-spacing="100"
    fixed="false"
    maxlength="200"
    placeholder=""
    placeholder-class="placeholder"
    placeholder-style=""
    selection-end="-1"
    selection-start="-1"
    show-confirm-bar="false"
  />
  <view
    class="at-textarea__counter"
  >
    0 / 200
  </view>
</view>
`;
